ASoC: SOF: Intel: use NHLT information for ES83x6 devices

plbossart opened the pull request:
We have to figure out the presence of DMICs and which SSP is used without DMI quirks, otherwise we'll drown in support requests.

Completely untested since I don't have hardware...

@hli25 could you take this for a spin and report what you find on GLK/TGL/JSL devices? There should be dev_dbg messages that you can enable by copying this file 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/7714253/sof-dyndbg.conf.txt) to /etc/modprobe.d/sof-dyndbg.conf
Enhancement tracked with https://github.com/thesofproject/linux/issues/3337

Support requests with active users in 
https://github.com/thesofproject/linux/issues/3336
https://github.com/thesofproject/linux/issues/3248

plbossart commented:
@kv2019i @juimonen this NHLT information could be used for BT offload and figure out which SSP is used if we ever support this for non-Chrome devices designed for Windows.

yangxiaohua2009 commented:
@plbossart Here is log  [glk_nhlt.log](https://github.com/thesofproject/linux/files/7725391/glk_nhlt.log) on glk. 
Currently we define SSP number in the machine driver and use three tplg for SSP0,1,2.
It would be better for the Intel platform driver to read SSP number from NHLT and choose the corresponding tplg file.

plbossart commented:
> @plbossart Here is log [glk_nhlt.log](https://github.com/thesofproject/linux/files/7725391/glk_nhlt.log) on glk. Currently we define SSP number in the machine driver and use three tplg for SSP0,1,2. It would be better for the Intel platform driver to read SSP number from NHLT and choose the corresponding tplg file.

Thanks @yangxiaohua2009 indeed that's the plan to autodetect the SSP, but in your example the mask is zero

[    8.746047] sof-audio-pci-intel-apl 0000:00:0e.0: intel_nhlt_get_dmic_geo: dmic number 0 max_ch 0
[    8.746054] sof-audio-pci-intel-apl 0000:00:0e.0: NHLT_LINK_SSP detected, ssp_mask 0x0

That makes no sense really, something is wrong here.
Can you please try to trace ssp_mask in intel_nhlt_ssp_endpoint_mask(), I wonder what happens here.

plbossart commented:
@yangxiaohua2009 I screwed-up with the initial code, please try the updated branch and let me know what you detect. thanks!

yangxiaohua2009 commented:
@plbossart [glk_2.log](https://github.com/thesofproject/linux/files/7731630/glk_2.log) Here's the new log.

plbossart commented:
@yangxiaohua2009 @hli25 can you take a look? This should remove the need for the SSP quirk and DMIC quirk. For the GPIO I have no solution.

yangxiaohua2009 commented:
@plbossart The kernel works well on glk [glk_111.log](https://github.com/thesofproject/linux/files/7844174/glk_111.log). However on CML it doesn't select sof in intel-dsp-config.c  [cml_111.log](https://github.com/thesofproject/linux/files/7844183/cml_111.log) and cannot recognize GPIO.
After force probing sof and GPIO, I see machine driver still using SSP2 instead of SSP0 on CML.

[cml_111_2.log](https://github.com/thesofproject/linux/files/7844227/cml_111_2.log)

plbossart commented:
Thanks for testing @yangxiaohua2009 

> @plbossart The kernel works well on glk [glk_111.log](https://github.com/thesofproject/linux/files/7844174/glk_111.log). However on CML it doesn't select sof in intel-dsp-config.c [cml_111.log]
(https://github.com/thesofproject/linux/files/7844183/cml_111.log) and cannot recognize GPIO. 

Sorry, I don't get how the SOF driver is not selected. I see two quirks for CML PCI IDs 0x06c8 and 0x02c8. Can you provide the result of 'cat /sys/bus/pci/devices/0000\:00\:1f.3/device'

Or did you take the patches and applied them on a different tree that's missing the CML quirks?

After force probing sof and GPIO, I see machine driver still using SSP2 instead of SSP0 on CML.
> 
> [cml_111_2.log](https://github.com/thesofproject/linux/files/7844227/cml_111_2.log)

The logs show SSP0 is used and passed to the machine driver? 

````
[    2.847963] sof-audio-pci-intel-cnl 0000:00:1f.3: NHLT_DEVICE_I2S detected, ssp_mask 0x1
````
 
Are you sure you didn't modify the quirk module parameter?

yangxiaohua2009 commented:
Since `!dmi_check_system(sof_es8336_quirk_table)` always return 1,  machine driver still uses SSP2. After removing it and making several changes the machine driver works [#24](https://github.com/plbossart/sound/pull/24). For SSP2 the i2s_link_mask is 5, for SSP1 i2s_link_mask is 3 and for SSP0 i2s_link_mask is 1 [apl_112.log](https://github.com/thesofproject/linux/files/7851794/apl_112.log).

yangxiaohua2009 commented:
Also, since the dsp driver already know which SSP to use from NHLT, can it select the corresponding tplg file? For example, on GLK I see OEMs using SSP2, SSP1 and SSP0.
For the GPIO, can the machine driver neglect PA control instead of returing error so that at least the user can get mic, headphone and HDMI working. @plbossart

yangxiaohua2009 commented:
'cat /sys/bus/pci/devices/0000:00:1f.3/device' returns 0x02c8. It's strange SOF is not used.

kv2019i reviewed (approved):
Code seems good to go.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -690,6 +690,25 @@ static int check_nhlt_dmic(struct snd_sof_dev *sdev)
+	struct nhlt_acpi_table *nhlt;
+	int ssp_mask = 0;
+
+	nhlt = intel_nhlt_init(sdev->dev);
```
the initialization can be moved to earlier when the nhlt is declared?

plbossart replied:
I try to avoid this when we have to test the results.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -690,6 +690,25 @@ static int check_nhlt_dmic(struct snd_sof_dev *sdev)
+
+		ssp_mask = intel_nhlt_ssp_endpoint_mask(nhlt, NHLT_DEVICE_I2S);
+
+		if (ssp_mask)
```
I would drop the blank lines.

plbossart replied:
yes, will change.

ujfalusi reviewed (changes requested):
Nice and clean series and it makes a lot of sense as well.
The only odd thing for me is that we are doing the SSP (I2S) connection query from NHLT in hda.c.

I had two minor comments as well.

plbossart commented:
> Nice and clean series and it makes a lot of sense as well. The only odd thing for me is that we are doing the SSP (I2S) connection query from NHLT in hda.c.

The naming is misleading indeed. HDA can mean
a) controller
b) link/bus
c) codec
d) dai / link DMA.

here hda.c is really about the controller initialization and the choice of a machine driver.

plbossart commented:
@yangxiaohua2009 Now I get the problem with dmi_check_system. Indeed this is wrong, because if we don't have a match we will ignore the NHLT information. It should be the other way around
a) first try to use the NHLT information
b) if there is a DMI match, override the NHLT stuff
c) if there is a user-specified quirk, override a) and b)

I am still trying to find out how to select the topology. One way to do this is as you mentioned by changing the topology file name, but that would not work if there is a quirk. I think we need some sort of means to override the SSP that's specified in the topology file.

For the GPIO, I am not sure what you are referring to, but we could use an optional GPIO that doesn't fail if it's not there. GPIOs are not my forte.

Thanks a lot of testing and your feedback, much appreciated.

plbossart commented:
@yangxiaohua2009 can you try the update, I tried to fix the quirk selection and add a mechanism to select the topology. compile-tested only since I don't have any hardware, let me know if this works for you.

plbossart commented:
The companion PR is here to generate all possible topology files: https://github.com/thesofproject/linux/pull/3338

tar file for testers to be installed in /lib/firmware/intel/sof-tplg (backup your setup first).
[es8336-topologies.tar.gz](https://github.com/thesofproject/linux/files/7858787/es8336-topologies.tar.gz)

yangxiaohua2009 commented:
[Here](https://github.com/thesofproject/linux/files/7860111/apl_113.log) is the result. I can see the suffix is not added.

i2s_link_mask is 1 (001) for ssp 0
i2s_link_mask is 3 (011) for ssp 1
i2s_link_mask is 5 (101) for ssp 2
So if BIT(0) is first checked, we always get SSP0.

Codec [here](https://github.com/plbossart/sound/compare/fix/sof-es8336-quirk...yangxiaohua2009:patch-2) works for me, and it also makes GPIO optional.

yangxiaohua2009 commented:
Do you need hardware? I may ask you Shanghai colleague to send you one. @plbossart

plbossart commented:
@yangxiaohua2009 the fundamental issue is that the NHLT reports two SSPs:

NHLT_DEVICE_I2S detected, ssp_mask 0x3

so there's really no way to know which of SSP0 or SSP1 is used. Can you share the ACPI tables (with acpidump) so that I can check the contents of the NHLT to double-check this point.

Edit: the suffix is not added because I made the assumption there is only one SSP reported by NHLT, and in case this is not true the suffix processing is skipped.

plbossart commented:
@yangxiaohua2009 I updated the branch to use SSP0 last, and also cleaned-up the machine driver. Let me know if this works better. I didn't take some of your changes for the optional gpio and I kept the error handling that you suggested be removed.

yangxiaohua2009 commented:
Tested on CML (SSP0) and APL (SSP1). The HDA driver is using tplg-ssp1 for CML and tplg-ssp2 for APL
[apl_115.log](https://github.com/thesofproject/linux/files/7868024/apl_115.log)
[cml_115.log](https://github.com/thesofproject/linux/files/7868025/cml_115.log)
@plbossart

plbossart commented:
@yangxiaohua2009 it seems my parsing of ssp mask is completely wrong. 

On CML the mask is 
NHLT_DEVICE_I2S detected, ssp_mask 0x1
but we use SSP1 in the file name.
loading topology:intel/sof-tplg/sof-cml-es8336-ssp1.tplg

on AP< the mask is 
NHLT_DEVICE_I2S detected, ssp_mask 0x3
but we use SSP2 in the file name
loading topology:intel/sof-tplg/sof-apl-es8336-ssp2.tplg

Darn it, it seems to be an off-by-one offset, fls does not return what i expected:
Note fls(0) = 0, fls(1) = 1, fls(0x80000000) = 32.

SSPs are zero based, so I should use fls(ssp_link_mask) - 1.

I also need to generate the SSP1 topologies since it's obviously needed in some cases. Gah.

plbossart commented:
@yangxiaohua2009 I updated the branch to correct the off-by-one and here are new topologies for the SSP1 case. Thanks again for your tests, much appreciated.

[es8336-topologies-2.tar.gz](https://github.com/thesofproject/linux/files/7871608/es8336-topologies-2.tar.gz)

yangxiaohua2009 commented:
The lastest code works well. The log.
[syslog.zip](https://github.com/thesofproject/linux/files/7880864/syslog.zip)

plbossart commented:
> The lastest code works well. The log. [syslog.zip](https://github.com/thesofproject/linux/files/7880864/syslog.zip)

Thank you @yangxiaohua2009 for your tests, much appreciated. I will leave this PR as is for now, I am still trying to figure out a way to use a generic topology and override the SSP used from the machine driver, but if I can find a way I'll send it upstream.

RezinoviyDed commented:
@plbossart Hello. I copy sof-dyndbg.conf.txt to /etc/modprobe.d/sof-dyndbg.conf and reboot my pc. My new ALSA information: http://alsa-project.org/db/?f=ff7db0a00552e54ad3aec3a7eedd701cfb57d90a
And $ cat /sys/module/snd_soc_sof_es8336/parameters/quirk
-1
Also i install new kernel - 5.16.1
Attached the [dmesg sof dsp nhlt output](https://github.com/thesofproject/linux/files/7898676/dmesg_sof_dsp_nhlt.txt) to the post.
Can it help?

plbossart commented:
> 

@RezinoviyDed you don't have the topology file for this device
[Ср янв 19 19:02:21 2022] sof-audio-pci-intel-icl 0000:00:1f.3: Direct firmware load for intel/sof-tplg/sof-jsl-es8336.tplg failed with error -2

I suggest you add the patches in this PR added on top of your kernel and retry with the topologies in this zip file. 
[es8336-topologies-2.tar.gz](https://github.com/thesofproject/linux/files/7898785/es8336-topologies-2.tar.gz)

RezinoviyDed commented:
I copied all tplg files from archive to /usr/lib/firmware/intel/sof-tplg/, but `sof-audio-pci-intel-icl 0000:00:1f.3: Direct firmware load for intel/sof-tplg/sof-jsl-es8336.tplg failed with error -2`

plbossart commented:
> I copied all tplg files from archive to /usr/lib/firmware/intel/sof-tplg/, but `sof-audio-pci-intel-icl 0000:00:1f.3: Direct firmware load for intel/sof-tplg/sof-jsl-es8336.tplg failed with error -2`

it's not helpful if you install the files from the topology zip and don't update the kernel.... The two go together.

RezinoviyDed commented:
Ok. I update the kernel. 
Before: 5.16.1-051601-generic
After: 5.16.2-051602-generic
I still see this warning((

plbossart commented:
> Ok. I update the kernel. Before: 5.16.1-051601-generic After: 5.16.2-051602-generic I still see this warning((

did you apply the patches on the updated kernel? If yes please provide the full dmesg after copying this file 
[sof-dyndbg.conf.txt](https://github.com/thesofproject/linux/files/7906040/sof-dyndbg.conf.txt) as /etc/modprobe.d/sof-dyndbg.conf and rebooting.

codesalatdev commented:
ES8336 CML here. Patched kernel `5.16.0-rc1` + your new tplgs. Dmesg log attached:
[dmesg.log](https://github.com/thesofproject/linux/files/7919557/dmesg.log)

plbossart commented:
> 

looks to have probed alright, do you still see a problem @codesalatdev ?

codesalatdev commented:
> > 
> 
> looks to have probed alright, do you still see a problem @codesalatdev ?

Well, I don't have sound, but topology-wise the selection looks good. The speakers make a short "pop" noise, which is the same result as manually picking the topology and using quirk 0x20

plbossart commented:
> Well, I don't have sound, but topology-wise the selection looks good. The speakers make a short "pop" noise, which is the same result as manually picking the topology and using quirk 0x20

You may need to unmute some controls with alsamixer, and alternatively test with the headphone to work-around possible issues with speaker enablement with GPIO.

codesalatdev commented:
> > Well, I don't have sound, but topology-wise the selection looks good. The speakers make a short "pop" noise, which is the same result as manually picking the topology and using quirk 0x20
> 
> You may need to unmute some controls with alsamixer, and alternatively test with the headphone to work-around possible issues with speaker enablement with GPIO.

hmm, I can't get this to work. I can mute and unmute in alsamixer, but neither speaker nor wired headphones work in any configuration. When I set speakers to 100% in alsamixer, I get nothing but white noise out of it. It's quiet, but it's there.

plbossart commented:
@yangxiaohua2009 can you try the update, I saw some reports of ESSX8326 ACPI HID being used, that updates *should* take care of it.

yangxiaohua2009 commented:
The change of ACPI HID works well. @plbossart
[cml_1_25.zip](https://github.com/thesofproject/linux/files/7931881/cml_1_25.zip)

yangxiaohua2009 commented:
Some ODMs use ssp2 for i2s data and ssp0 for mclk. Is mclk configuration set in NHLT? How is this situation handled in windows?

plbossart commented:
> Some ODMs use ssp2 for i2s data and ssp0 for mclk. Is mclk configuration set in NHLT? How is this situation handled in windows?

@yangxiaohua2009 your wording does not reflect how the hardware is built. The mclk is not tied to any specific SSP. There are two MCLK pins (mclk_id=0 and mclk_id=1), regardless of the number of SSPs. This can be controlled through the topology.

````
SSP_CONFIG_DATA(type, idx, valid bits, mclk_id)
 mclk_id is optional
````

It's not really correct in the sense that the mclk should be configured separately, but in practice it's not problematic.

by default we use mclk_id == 0 for all topologies, only a very small number of chromebooks used mclk_id = 1

so the question is: do you know of any devices that use mclk_id == 1?

plbossart commented:
one more update to make kbuild bot happy when some platforms are not selected.

````
sound/hda/intel-dsp-config.c:38:41: warning: unused variable 'essx_83x6' [-Wunused-const-variable]
````

yangxiaohua2009 commented:
> > Some ODMs use ssp2 for i2s data and ssp0 for mclk. Is mclk configuration set in NHLT? How is this situation handled in windows?
> 
> @yangxiaohua2009 your wording does not reflect how the hardware is built. The mclk is not tied to any specific SSP. There are two MCLK pins (mclk_id=0 and mclk_id=1), regardless of the number of SSPs. This can be controlled through the topology.
> 
> ```
> SSP_CONFIG_DATA(type, idx, valid bits, mclk_id)
>  mclk_id is optional
> ```
> 
> It's not really correct in the sense that the mclk should be configured separately, but in practice it's not problematic.
> 
> by default we use mclk_id == 0 for all topologies, only a very small number of chromebooks used mclk_id = 1
> 
> so the question is: do you know of any devices that use mclk_id == 1?

Yes 90% of all the devices use mclk_id == 0, and the rest use mclk_id == 1. Those who don't have sound with correct SSP setting may need to try the other tplg with different mclk_id.

libinyang commented on sound/soc/sof/intel/hda.c:
```diff
@@ -607,24 +605,35 @@ static int hda_init(struct snd_sof_dev *sdev)
 
-	return 0;
+	/* allow for module parameter override */
+	if (dmic_num_override != -1) {
```
As we also check the module parameter, maybe it is better to change the function name to check_dmic_num() to avoid confusion?

plbossart replied:
good suggestion, will add. Thanks @libinyang

plbossart replied:
done

libinyang commented on sound/soc/sof/intel/hda.c:
```diff
@@ -632,6 +632,23 @@ static int check_nhlt_dmic(struct snd_sof_dev *sdev)
+		if (ssp_mask)
+			dev_info(sdev->dev, "NHLT_DEVICE_I2S detected, ssp_mask %#x\n", ssp_mask);
+
+		intel_nhlt_free(nhlt);
```
intel_nhlt_free() won't be called if intel_nhlt_has_endpoint_type(nhlt, NHLT_LINK_SSP) returns 0.
Maybe we can:
```
nhlt = intel_nhlt_init(sdev->dev);
if (!nhlt)
    return ssp_mask;

ssp_mask = intel_nhlt_ssp_endpoint_mask(nhlt, NHLT_DEVICE_I2S);
if (ssp_mask)
    dev_info(sdev->dev, "NHLT_DEVICE_I2S detected, ssp_mask %#x\n", ssp_mask);
intel_nhlt_free(nhlt);

return ssp_mask;
````

plbossart replied:
good catch @libinyang. This needs to be modified indeed. Thanks for the feedback, much appreciated.

plbossart replied:
done

libinyang commented on sound/soc/intel/boards/sof_es8336.c:
```diff
@@ -483,6 +505,9 @@ static int sof_es8336_probe(struct platform_device *pdev)
 
+	if (quirk & SOF_ES8336_ENABLE_DMIC)
+		dmic_be_num = 2;
+
```
In the original code, dmic_be_num is set before quirk_override. This mean the quirk_override won't impact on the dmic setting. Now dmic setting will be changed if quirk_override is set. Personally, I like this patch behavior more if there are not products impacted.

plbossart replied:
yes it's intentional that the quirk override will trump NHTL information and DMI.
Of course people could screw up by adding mics in the HDA quirk and removing them in the machine driver quirk. That would lead to an inconsistency between machine driver and topology, but that's an enforced error.

plbossart commented:
git merge issue, pushed update to fix compilation errors which didn't exist in my branch...

plbossart commented:
and another push for arm64 compilation

yangxiaohua2009 commented:
I compiled with the fix [5929fad](https://github.com/thesofproject/linux/pull/3338/commits/5929fad3b923ae84b691e4444e006e0ea2e0d5a6), the sof-essx8336 is loaded but the ES8326 codec driver still cannot be probed. The codec driver can be probed in kernel version 5.11 on the same machine though. I add dyndbg=drivers/base/dd.c in grub and find the probing of all i2c drivers including the codec driver is deferred. 
I installed the kernel on another ES8336 machine and the ES8336 codec driver is probed successfully. 
Here are the logs, two success and one failure [83x6_tgl_2_14.zip](https://github.com/thesofproject/linux/files/8059230/83x6_tgl_2_14.zip). @plbossart

plbossart commented:
You can debug this with blacklist and manual modprobe @yangxiaohua2009 

add
blacklist snd-soc-es8326
blacklist snd-soc-sof-es8336

boot and then do a 
modprobe snd-soc-es8326
If you can't probe successfully then it's your driver that has a missing dependency

Then when this has probed, so 
modprobe snd-soc-sof-es8336

yangxiaohua2009 commented:
I tried your suggestion, the result is the same:
`sudo modprobe snd-soc-es8326`
[ +25.212922] bus: 'i2c': __driver_probe_device: matched device i2c-ESSX8326:00 with driver es8326
[  +0.017923] bus: 'i2c': really_probe: probing driver es8326 with device i2c-ESSX8326:00
[  +0.002284] es8326 i2c-ESSX8326:00: Driver es8326 requests probe deferral
[  +0.000015] i2c i2c-ESSX8326:00: Added to deferred list
`sudo modprobe snd-soc-sof-es8336`
[2月15 09:25] bus: 'platform': __driver_probe_device: matched device sof-essx8336 with driver sof-essx8336
[  +0.000005] bus: 'platform': really_probe: probing driver sof-essx8336 with device sof-essx8336
[  +0.000006] sof-essx8336 sof-essx8336: quirk mask 0x0
[  +0.000002] sof-essx8336 sof-essx8336: quirk SSP0
[  +0.000744] sof-essx8336 sof-essx8336: Driver sof-essx8336 requests probe deferral
[  +0.000013] platform sof-essx8336: Added to deferred list

I tried to modify the codec driver and the dailink in sof-essx8336, there is no effect, which means the problem is not related to them.
I modified the soc-acpi-intel-tgl-match.c, using ESSX8326 as id and the result is the same.
Also, I found that the probe of all i2c drivers are deferred, including i2c_hid_acpi.
I guess the modification in sound/soc/sof/intel/hda.c affects pci/acpi ?

RezinoviyDed commented:
I install this kernel: https://github.com/yangxiaohua2009/custom-kernel and copy all tplg from es8336-topologies-2.tar.gz to sof-tplg. The messages in the dmesg have changed.
Attached [dmesg.log](https://github.com/thesofproject/linux/files/8066888/dmesg.log)

yangxiaohua2009 commented:
> I install this kernel: https://github.com/yangxiaohua2009/custom-kernel and copy all tplg from es8336-topologies-2.tar.gz to sof-tplg. The messages in the dmesg have changed. Attached [dmesg.log](https://github.com/thesofproject/linux/files/8066888/dmesg.log)

I guess you are using a jsl device. Unzip and copy this ri to /lib/firmware/intel/sof/
[sof-jsl.zip](https://github.com/thesofproject/linux/files/8066909/sof-jsl.zip)

RezinoviyDed commented:
Ok, i copy ri file to intel/sof.
Attached [dmesg.log](https://github.com/thesofproject/linux/files/8067043/dmesg.log)

My ALSA information: http://alsa-project.org/db/?f=096a36c0af2478b62422496e6d5b633974886b90

yangxiaohua2009 commented:
> Ok, i copy ri file to intel/sof. Attached [dmesg.log](https://github.com/thesofproject/linux/files/8067043/dmesg.log)
> 
> My ALSA information: http://alsa-project.org/db/?f=096a36c0af2478b62422496e6d5b633974886b90

You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully. Need @plbossart to figure out why.

plbossart commented:
> > Ok, i copy ri file to intel/sof. Attached [dmesg.log](https://github.com/thesofproject/linux/files/8067043/dmesg.log)
> > My ALSA information: http://alsa-project.org/db/?f=096a36c0af2478b62422496e6d5b633974886b90
> 
> You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully. Need @plbossart to figure out why.

I have no idea @yangxiaohua2009. the tests show that devices with I2C codecs such as RT5682 and DA7219 work fine.

plbossart commented:
updated branch to cover CNL as reported by @realnickel

updated topologies are here: 
[es8336-topologies-3.tar.gz](https://github.com/thesofproject/linux/files/8076326/es8336-topologies-3.tar.gz)

plbossart commented:
> > > Ok, i copy ri file to intel/sof. Attached [dmesg.log](https://github.com/thesofproject/linux/files/8067043/dmesg.log)
> > > My ALSA information: http://alsa-project.org/db/?f=096a36c0af2478b62422496e6d5b633974886b90
> > 
> > 
> > You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully. Need @plbossart to figure out why.
> 
> I have no idea @yangxiaohua2009. the tests show that devices with I2C codecs such as RT5682 and DA7219 work fine.

@yangxiaohua2009 I just tested this PR on a JasperLake device, works fine, see log 
[dmesg.log](https://github.com/thesofproject/linux/files/8076602/dmesg.log)

I think it's your kernel version that has a problem, this PR looks just fine.

yangxiaohua2009 commented:
> > > > Ok, i copy ri file to intel/sof. Attached [dmesg.log](https://github.com/thesofproject/linux/files/8067043/dmesg.log)
> > > > My ALSA information: http://alsa-project.org/db/?f=096a36c0af2478b62422496e6d5b633974886b90
> > > 
> > > 
> > > You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully. Need @plbossart to figure out why.
> > 
> > 
> > I have no idea @yangxiaohua2009. the tests show that devices with I2C codecs such as RT5682 and DA7219 work fine.
> 
> @yangxiaohua2009 I just tested this PR on a JasperLake device, works fine, see log [dmesg.log](https://github.com/thesofproject/linux/files/8076602/dmesg.log)
> 
> I think it's your kernel version that has a problem, this PR looks just fine.

Can you share me the defconfig file you're using? Maybe the .config file I use lack some Kconfig.
I will tell the ODMs to compile with this PR. Thank you.

luco5826 commented on sound/hda/intel-dsp-config.c:
```diff
@@ -414,8 +425,13 @@ static const struct config_entry *snd_intel_dsp_find_config
 			continue;
-		if (table->codec_hid[0] && !acpi_dev_present(table->codec_hid, NULL, -1))
-			continue;
+		if (table->codec_hid) {
+			int i;
+
+			for (i = 0; i < table->codec_hid->num_codecs; i++)
+				if (!acpi_dev_present(table->codec_hid->codecs[i], NULL, -1))
+					continue;
+		}
```
Is this snippet equivalent to the previous version? 
In this case the `continue` keyword only skips the inner for loop (the one at line 431) while it probably should "continue" the outer loop, or am I missing something?

plbossart replied:
great feedback @luco5826, this is completely wrong indeed. will fix.

plbossart commented:
> Can you share me the defconfig file you're using? Maybe the .config file I use lack some Kconfig. I will tell the ODMs to compile with this PR. Thank you.

We are using minimal configurations to compile quickly and only test audio.
[work-defconfig.txt](https://github.com/thesofproject/linux/files/8080650/work-defconfig.txt)

I am not sure this is suitable for ODMs, just sharing to show what I tested.

FWIW, you can know add the recommended SOF options on top of a distro config with https://github.com/thesofproject/kconfig/blob/master/kconfig-distro-sof-update.sh

darrell-k commented:
> 



> updated branch to cover CNL as reported by @realnickel
> 
> updated topologies are here: [es8336-topologies-3.tar.gz](https://github.com/thesofproject/linux/files/8076326/es8336-topologies-3.tar.gz)

Continuing from https://github.com/thesofproject/linux/issues/3248 

I've installed the latest kernel from this pr, and the above topologies. No manual quirks added. 

Still no sound from internal speakers, but headset output jack detection works and sound is available after messing around with alsamixer. 

Internal mic not detected and headset mic jack detection does not work (still shows as "unplugged" after plugging). 

Where do we go from here?

dengshengliu commented:
I’ve installed the latest kernel(5.16.0-rc1+) on GLK-ES8336#2 platform. the
Sound card works fine, 
However, I modified the code of es8316.c, and couldn't find the sound card after compilation. I changed it to the previous one again, but still couldn't find it，why？
![Screenshot from 2022-02-20 22-28-51](https://user-images.githubusercontent.com/43395497/154898021-54d4f31b-ff96-435c-8290-673e9d8ee06d.png)
[dmesg_220221.txt](https://github.com/thesofproject/linux/files/8106617/dmesg_220221.txt)

yangxiaohua2009 commented:
> I’ve installed the latest kernel(5.16.0-rc1+) on GLK-ES8336#2 platform. the Sound card works fine, However, I modified the code of es8316.c, and couldn't find the sound card after compilation. I changed it to the previous one again, but still couldn't find it，why？ ![Screenshot from 2022-02-20 22-28-51](https://user-images.githubusercontent.com/43395497/154898021-54d4f31b-ff96-435c-8290-673e9d8ee06d.png) [dmesg_220221.txt](https://github.com/thesofproject/linux/files/8106617/dmesg_220221.txt)

Are you using HDA in BIOS? Changing codec driver doesn't affect your machine driver probe.

dengshengliu commented:
Enter the specified kernel(5.16.0-rc1+) system and restart the computer,
 the system will kernel panic
![a526149818911954ad77faf530858fe](https://user-images.githubusercontent.com/43395497/154899960-587cb91c-8b11-4767-aa95-bc8c68b7261a.jpg)
.

dengshengliu commented:
Those that use the everest-semi codec are not HDA types.

yangxiaohua2009 commented:
> Those that use the everest-semi codec are not HDA types.

You may use `git reset --hard` to make sure that your source code is the same as the origin.

RezinoviyDed commented:
> You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully.

@yangxiaohua2009 Did you manage to solve this problem?

yangxiaohua2009 commented:
> > You are facing the same problem as mine. The ESSX8336 i2c driver is not probed successfully.
> 
> @yangxiaohua2009 Did you manage to solve this problem?

Nop. You may use the previous kernel [here](https://github.com/yangxiaohua2009/custom-kernel) and see what happens.

darrell-k commented:
> Enter the specified kernel(5.16.0-rc1+) system and restart the computer, the system will kernel panic ![a526149818911954ad77faf530858fe](https://user-images.githubusercontent.com/43395497/154899960-587cb91c-8b11-4767-aa95-bc8c68b7261a.jpg) .

I got this until installing the latest topologies [es8336-topologies-3.tar.gz](https://github.com/thesofproject/linux/files/8076326/es8336-topologies-3.tar.gz)

dengshengliu commented:
Those that use the everest-semi codec are not HDA types.

You may use git reset --hard to make sure that your source code is the same as the origin.

==>
do it your way and it still won't work, i re-git the new code from gitHub, 
after the installation, the sound card works fine.



![Screenshot from 2022-02-21 20-54-26](https://user-images.githubusercontent.com/43395497/155056263-0dcca283-6cc5-4ec6-8bce-5606eb4d71f6.png)

plbossart commented:
> > Enter the specified kernel(5.16.0-rc1+) system and restart the computer, the system will kernel panic ![a526149818911954ad77faf530858fe](https://user-images.githubusercontent.com/43395497/154899960-587cb91c-8b11-4767-aa95-bc8c68b7261a.jpg) .

It's been reported before and I don't quite see what happens here. This is a bog-standard HDMI handling, it's been tried and tested for several years already.

Not having a local device makes it hard for me to see such issues.

plbossart commented:
@ujfalusi @yangxiaohua2009 @bardliao @ranj063 Can I bother to look at this PR, I'd like to send this set upstream. This should solve most problems with 5.18. We'll have to refine for 5.19 the mclk id support and es8236 support.

bardliao commented on sound/hda/intel-dsp-config.c:
```diff
@@ -424,8 +424,10 @@ static const struct config_entry *snd_intel_dsp_find_config
+				if (acpi_dev_present(table->codec_hid->codecs[i], NULL, -1))
+					break;
+			if (i == table->codec_hid->num_codecs)
+				continue;
```
@plbossart Should this be squashed into the previous commit?

plbossart replied:
yes indeed, nice catch, I squashed a fixup in the wrong file. Thanks for spotting this @bardliao, fixed now.

fredoh9 commented:
SOFCI TEST

plbossart commented:
@ujfalusi can I please ask you to take a look, you had initial objections and I don't want to merge without your feedback. thanks!

plbossart commented:
https://sof-ci.01.org/linuxpr/PR3338/build7261/devicetest/ had CI issues, let's redo the tests

plbossart commented:
SOFCI TEST

ujfalusi commented on sound/hda/intel-dsp-config.c:
```diff
@@ -31,7 +32,12 @@ struct config_entry {
 	u8 acpi_hid[ACPI_ID_LEN];
 	const struct dmi_system_id *dmi_table;
-	u8 codec_hid[ACPI_ID_LEN];
+	const struct snd_soc_acpi_codecs *codec_hid;
```
@plbossart, out of curiosity, what 'hid' stands for? Can this be simply named as 'codecs'?

plbossart replied:
It's an ACPI acronym for 'Hardware ID'. There's also 'Compatible ID'.
this is represented in DSDT, e.g.

````      
Device (ESSX)
                {
                    Name (_ADR, Zero)  // _ADR: Address
                    Name (_HID, "ESSX8336")  // _HID: Hardware ID
                    Name (_CID, "ESSX8336")  // _CID: Compatible ID
                    Name (_DDN, "ES8336")  // _DDN: DOS Device Name
````

ujfalusi reviewed (approved):
Looks fine to my eyes.
